### PR TITLE
Draft: Database Redesign

### PR DESCRIPTION
**Pull request by Hitissum**
image\
[*Link to Diagram with DBML*](https://dbdiagram.io/d/624b40c6d043196e39f34471)

I wanted to go with a system that would be easy to add new conversation specific user information in the future, things like whether or not they are an administrator of the conversation or conversation specific usernames. 

I really like the benefit this has when it comes to conversations that are multi-user or one-on-one as it really simplifies how much you have to do to successfully implement the feature, rather than maintaining two distinct features with different tables you can work on and maintain one large one.

## **What additional changes are needed to implement this feature**

Unfortunately one of the trade offs of making such a large change is going to be it's going to require quite a bit of non-trivial work up front to pull off but when weighted against the detriments of sticking with an inflexible model is well worth it. 

Some of the work that's going to be required
- Anything that utilizes the prior model will need to be worked on to fit the new system.
- A user list for non-direct communication will need to be implemented
- A system for inviting users to conversations to a conversation
  - Perhaps inviting the users opens the conversation with them but a prompt asking if they'd like to actually join is presented.

Obviously there would need to be a meetings to discuss what features should come along with this so everything is polished before it's shipped to production. 

## **If our app were currently deployed, what steps would we need to take to move to this new feature without disrupting service**

Well, obviously there would be database migration to be worried about luckily the library we use for interacting with our DB [has built in migration support](https://sequelize.org/docs/v6/other-topics/migrations/).  

In terms of worrying about legacy clients or applications written to interface with our platform the API could be versioned as to give those who need it time to migrate their applications with the added benefit of anyone who hasn't updated their client isn't suddenly broken one day. 

Special considerations would be need to made for situations where one or multiple within a direct conversation as introducing extra users to the conversation would most likely cause unintended behavior.

## **Other ways this could have been done**
Adding separate tables for group conversations does have the benefit of making it easier to support legacy applications but would mean you wouldn't be reaping the reward of having an easily extensible member table. It would be need to be readdressed at a later point when the normal conversation model would need support for user specific data.

### **Finally my proposed DB schema as viewed through LibreOffice Base**
image
*Although the seeding understandably errors, it correctly syncs the model to the DB*

### CODE REVIEW

File: server/db/models/index.js [hunk 1]
```diff
@@ -1,17 +1,25 @@
...
+Member.belongsTo(User);
+
+Conversation.hasMany(Member);
+Member.belongsTo(Conversation);
+
 Conversation.hasMany(Message);
+Message.belongsTo(Conversation);
+
+Member.hasMany(Message);
+Message.belongsTo(Member);
```
Comment (cgarrovillo) [inline server/db/models/index.js:18]: This is good in principle slightly_smiling_face
There's just a small mistake in the Sequelize implementation. 
```
User.belongsToMany(Conversation, {through: Member});
Conversation.belongsToMany(User, {through: Member});
```
This is the correct way of implementing a many-to-many relationship in Sequelize.

**Review by cgarrovillo — COMMENTED** [review]
Thank you for submitting this PR!

- Could you provide more details about the socket code,  route handlers,  frontend UI and React state?
- Can you provide a step-by-step migration plan, such that service is not disrupted to users?

Let us know when this PR is ready for a second review!

**Comment by Hitissum** [comment]
### **Could you provide more details about the socket code, route handlers, frontend UI and React state?**

### _Sockets_

The `new-message` packet is generally insecure and broadcasts anything its sent to everyone. Sockets should be assigned [socket.io rooms](https://socket.io/docs/v3/rooms/) based on the session cookie given on connection. Getting rid of the redundancy and making it so `POST /messages/` also sends a message to the correct rooms would be ideal and eliminate arbitrary information dispatching.

### _Routes_

A lot of the backend relies on searching for `user1Id` and `user2Id` when the conversation isn't known which wouldn't work for our new system. So it would mean the need to modify both `POST /messages/` and `GET /conversations/` to not use them as fallback values when a conversationId is not known.

### _Frontend UI_

With the new conversation model the UI will have to handle the conversation title and icon differently depending on whether or not the conversation is a direct one or not. The main difference being that a direct conversation should show the other user's profile picture and name and the group conversation should have a title and icon that's designated by it's creator.

Creating a conversation that you can invite people to and a direct one will be different the basic premise of a direct one will be the same but a arbitrary user one will be created by the user who is then instructed to select friends they would like to participate in it. So a menu for that would need to be introduced. Arbitrary user conversations will also need a member list on the right to indicate who is within the conversation as well.

### _React state_

One of the biggest changes that comes to mind is `activeConversation` in it's current state it takes the name of the current user you are conversing with which would need to be substituted with something more suitable like the ID or in the case of a conversation that hasn't been created yet a temporary placeholder.

The basis of users is an interesting one on the platform with each of them having their own conversation within `conversations` which doesn't make sense within the new system. This would mean the search bar which utilizes the old principal would need to be refactored to support searching both conversations the session is in and users that are within the system.

### **Can you provide a step-by-step migration plan, such that service is not disrupted to users?**

Making such a large change to how our application fundamentally works comes with the cost of making the deployment a lot more convoluted as a result. However, I believe it is possible to do so due to our support for legacy clients and the simplicity of our application.

1. Prior to deployment we will roll a secondary database with a carbon copy of our production database up to that point
2. Deploy an interim build of our new backend which serves two purposes
   - First, utilizing [sequelize migrations](https://sequelize.org/docs/v6/other-topics/migrations/) bring our copy to the newest schema
   - Secondly, listen to and hold onto mirrored web traffic being sent [ from our nginx configuration](https://faun.pub/mirroring-incoming-web-traffic-with-nginx-fit-devops-d688ddca7d30)
3. After the migration is complete on the copy replay the requests we held onto for our backwards compatible endpoints
4. Swap the roles of both servers if it is deemed it is healthy enough to handle it

Because our endpoints are all deterministic and don't require any randomness in their usage it would work with the only potential issue being that the updatedAt timestamps would be off by a little bit at most. Although if it were demanded the interim build could pass the cached request timestamp along.

File: server/db/models/index.js [hunk 2]
```diff
@@ -1,17 +1,22 @@
...
 const Message = require("./message");
 
 // associations
 
-User.hasMany(Conversation);
-Conversation.belongsTo(User, { as: "user1" });
-Conversation.belongsTo(User, { as: "user2" });
-Message.belongsTo(Conversation);
+Conversation.belongsToMany(User, {through: Member});
+User.belongsToMany(Conversation, {through: Member});
```
Comment (uimarshall) [inline server/db/models/index.js:9]: 💯

**Review by uimarshall — APPROVED** [review]
Thank you for your work on this PR.
I like the way you go about addressing the comments, but the Siquelize migrations you mentioned are not meant for database migrations but to [actually create a table in the database](https://sequelize.org/docs/v6/other-topics/migrations/#:~:text=Running%20Migrations%E2%80%8B,run%20db%3Amigrate%20command.). 
Consider this PR approved. Please do not merge this PR since it is only a draft. 👍

**Pull request closed without merging**